### ISSUE TITLE
[ENH] grid/random search tuning for time series classification

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#2116]
We should ensure one of two things:
* that grid/random search tuning is straighforward for TSC with `sklearn` and onboard tools.
* that a grid/random search tuning wrapper is written for TSC, similar to `ForecastingGridSearch`

In either case, this should be added to the TSC tutorial.

@TonyBagnall, @MatthewMiddlehurst, @ABostrom, @chrisholder, am I right in assuming this does not exist yet?
If it does, kindly correct me, and I would appreciate pointers (and @jmwhyte possibly too).

**Comment by jmwhyte** [comment]
@fkiraly @TonyBagnall @MatthewMiddlehurst @ABostrom @chrisholder I will be very happy to have pointers, and ideally, a clear tutorial on the hyper-parameter tuning and cross-validation tasks in TSC.

**Comment by MatthewMiddlehurst** [comment]
Ideally, they should just work with the sklearn functions? There isn't any fundamental difference between cross-validating/tuning time series data and regular data.

Are there any cases where this does not work currently that you are aware of? We use this sklearn function fine when running our experiments.

`train_probs = cross_val_predict(classifier, X_train, y=y_train, cv=cv_size, method="predict_proba")`

**Comment by MatthewMiddlehurst** [comment]
There is still a classification notebook. 

https://github.com/alan-turing-institute/sktime/blob/main/examples/02_classification.ipynb

**Comment by TonyBagnall** [comment]
all classifiers should work with scikit cross validation, although admittedly I have not fully tested this. I  plan to greatly expand the classification notebook in due course, but I can add a section on this sooner than the whole thing. Structurally, it will be identical to scikit learn, but it would be good to present what parameters there are for each algorithm.

**Comment by TonyBagnall** [comment]
dont really see the need for a wrapper, unless you mean a TunedClassifier wrapper such as that we use in tsml https://github.com/uea-machine-learning/tsml/blob/master/src/main/java/machine_learning/classifiers/tuned/TunedClassifier.java

**Comment by fkiraly** [comment]
> all classifiers should work with scikit cross validation, although admittedly I have not fully tested this. I plan to greatly expand the classification notebook in due course, but I can add a section on this sooner than the whole thing. Structurally, it will be identical to scikit learn, but it would be good to present what parameters there are for each algorithm.

Are we absolutely certain you can use `sklearn` grid search with `sktime` time series classifiers, and it doesn't jumble the different series together into train/test set?

I'd consider the minimum solution to this a section in the classifier notebook that shows how to do this properly, otherwise users will not know how to do it.

For instance, I also couldn't tell anyone how to do proper grid/random search with the TSC module!

**Comment by MatthewMiddlehurst** [comment]
I don't believe anyone has tested the suite of sklearn tuners/cross-validators using the sktime data formats used for TSC. I don't anticipate there being any issue with the classifiers themselves, rather whether the sklearn functions can accept a 3d numpy array, nested dataframe or dataframe with unequal length series.

I personally have not run into any issues, but I haven't looked too deeply into it. At the end of the day, these functions won't have to look at the actual data at all, at most they need to know how many cases are in the input data and be able to move/split the cases. If they are somehow swapping multivariate dimensions around without crashing, they are doing some really funky stuff in those functions.

I agree we should figure out what is available from sklearn and test all of our accepted data formats for TSC on them. Perhaps add a few tests.

**Comment by fkiraly** [comment]
> I agree we should figure out what is available from sklearn and test all of our accepted data formats for TSC on them. Perhaps add a few tests.

Yes, plus write a tutorial. I can conceive that `numpy3D` and `nested_univ` input can possibly work, because `train_test_split` works on the same index and doesn't care about the rest, but a `sklearn` wrapper will probably not accept any other of our internal data formats.

**Comment by TonyBagnall** [comment]
well given we do not yet have a single use case for anything that is not numpy or nested_univ, scikit learn compatibility is sufficient for me. It you want to do write a wrapper, please do, I do not see the immediate need for it. I am working on extending the notebooks, and this issue has moved it up my order of work, I would like to verify compatibility with scikit learn tools for our core classifiers

**Comment by fkiraly** [comment]
> scikit learn compatibility is sufficient for me

We have not verified that yet.

**Comment by MatthewMiddlehurst** [comment]
We now have a number of tests for verifying compatibility between the sklearn model selection methods and the sktime data formats/classification interface.

I think as long as these remain compatible, there is no need for bespoke sktime implementations.

**Comment by fkiraly** [comment]
@MatthewMiddlehurst, does this work for all `sktime` compatible input types, e.g., nested `pd.DataFrame`?

**Comment by MatthewMiddlehurst** [comment]
There are tests are for both 3d numpy and nested DataFrame.

**Comment by MatthewMiddlehurst** [comment]
Is this (and #2132) good? I don't think it is necessary to test every single method on every single classifier. Just testing the interface and our input data types should be enough.

**Comment by fkiraly** [comment]
@MatthewMiddlehurst, does the grid search work for all `sktime` input types?

**Comment by MatthewMiddlehurst** [comment]
There are tests for both 3d numpy and nested DataFrames which pass. 2d numpy and dataframes should already be accepted sklearn inputs.

**Comment by MatthewMiddlehurst** [comment]
I'm pretty sure the only portion remaining from the original issue is adding sklearn cv/tuning to the TSC tutorial.

**Comment by jmwhyte** [comment]
@MatthewMiddlehurst @fkiraly  I will be very keen to see these additions to the tutorial!

**Comment by fkiraly** [comment]
> I'm pretty sure the only portion remaining from the original issue is adding sklearn cv/tuning to the TSC tutorial.

Well, let's keep it upen then until it's added by some kind volunteer 😄

**Comment by jmwhyte** [comment]
Ok, content coming soon I hope. (I just have to find time to read through the various docs about contributing, and tidy the code when I get a chance.)
